File: LC/Medium/199.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        ans = []
        q = [root]
        
        while q:
            last_val = None
            
            for _ in range(len(q)):
                curr = q.pop(0)
                
                if curr:
                    last_val = curr.val
                
                    if curr.left:
                        q.append(curr.left)
                    if curr.right:
                        q.append(curr.right)
            
            if last_val != None:
                ans.append(last_val)

        return ans

File: LC/C++/Medium/199.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        std::vector<int> ans{};
        std::queue<TreeNode*> q{};
        
        if (root) q.push(root);

        while (!q.empty()) {
            int last = -1;
            size_t n = q.size();

            while (n-- > 0) {
                TreeNode* curr = q.front();
                q.pop();

                last = curr->val;

                if (curr->left) q.push(curr->left);
                if (curr->right) q.push(curr->right);
            }

            ans.push_back(last);
        }

        return ans;
    }
};